compact_test.go

```diff
@@ -154,153 +157,231 @@ func TestNoPanicFor0Tombstones(t *testing.T) {
-					NumSeries:     10,
-					NumTombstones: 3,
-				}),
-				metaRange("2", 540, 560, nil),
+				metaRange("7", 1522087200000, 1522094400000, nil),
+				metaRange("9", 1522094400000, 1522101600000, nil),
+				metaRange("10", 1522051200000, 1522087200000, nil), // 10h
+				metaRange("11", 1522101600000, 1522108800000, nil),
+				metaRange("12", 1522108800000, 1522116000000, nil),
 			},
-			expected: []string{"1"},
+			expected: []string{"7", "9", "11"},
 		},
-		// For small blocks, do not compact tombstones, even when fresh appears.
 		{
+			// ts=2018-03-27T09:00:00.034468506Z
 			metas: []dirMeta{
-				metaRange("1", 0, 60, &BlockStats{
-					NumSeries:     10,
-					NumTombstones: 3,
-				}),
-				metaRange("2", 60, 80, nil),
+				metaRange("10", 1522051200000, 1522087200000, nil), // 10h
+				metaRange("12", 1522108800000, 1522116000000, nil),
+				metaRange("13", 1522087200000, 1522108800000, nil), // 6h
+				metaRange("14", 1522116000000, 1522123200000, nil),
+				metaRange("15", 1522123200000, 1522130400000, nil),
+				metaRange("16", 1522130400000, 1522137600000, nil),
+				// -blackbox-exporter stuff (unrelated?)
 			},
-			expected: nil,
+			expected: []string{"12", "14", "15"},
 		},
-		// Regression test: we were stuck in a compact loop where we always recompacted
-		// the same block when tombstones and series counts were zero.
 		{
+			// ts=2018-03-27T15:00:00.012415816Z
 			metas: []dirMeta{
-				metaRange("1", 0, 540, &BlockStats{
-					NumSeries:     0,
-					NumTombstones: 0,
-				}),
-				metaRange("2", 540, 560, nil),
+				metaRange("10", 1522051200000, 1522087200000, nil), // 10h
+				metaRange("13", 1522087200000, 1522108800000, nil), // 6h
+				metaRange("16", 1522130400000, 1522137600000, nil),
+				metaRange("17", 1522108800000, 1522130400000, nil), // 6h
+				metaRange("18", 1522137600000, 1522144800000, nil),
+				metaRange("19", 1522144800000, 1522152000000, nil),
+				metaRange("20", 1522152000000, 1522159200000, nil),
 			},
-			expected: nil,
+			expected: []string{"16", "18", "19"},
 		},
-		// Regression test: we were wrongly assuming that new block is fresh from WAL when its ULID is newest.
-		// We need to actually look on max time instead.
-		//
-		// With previous, wrong approach "8" block was ignored, so we were wrongly compacting 5 and 7 and introducing
-		// block overlaps.
 		{
+			// ts=2018-03-27T15:00:05.769323478Z
 			metas: []dirMeta{
-				metaRange("5", 0, 360, nil),
-				metaRange("6", 540, 560, nil), // Fresh one.
-				metaRange("7", 360, 420, nil),
-				metaRange("8", 420, 540, nil),
+				metaRange("10", 1522051200000, 1522087200000, nil), // 10h
+				metaRange("13", 1522087200000, 1522108800000, nil), // 6h
+				metaRange("17", 1522108800000, 1522130400000, nil), // 6h
+				metaRange("20", 1522152000000, 1522159200000, nil),
+				metaRange("21", 1522130400000, 1522152000000, nil), // 6h
 			},
-			expected: []string{"7", "8"},
+			expected: []string{"13", "17", "21"},
```

Comment:
This is important part proving that 2.2.1 works locally, but not for @dswarbrick